pkg/gardenlet/controller/backupbucket/reconciler.go

```diff
@@ -0,0 +1,441 @@
+			return reconcile.Result{}, fmt.Errorf("failed to add finalizer: %w", err)
+		}
+	}
+
+	if updateErr := updateBackupBucketStatusProcessing(ctx, r.GardenClient, backupBucket, "Reconciliation of Backup Bucket state in progress.", 2, r.Clock); updateErr != nil {
+		return reconcile.Result{}, fmt.Errorf("could not update status after reconciliation start: %w", updateErr)
+	}
+
+	secret, err := kutil.GetSecretByReference(ctx, r.GardenClient, &backupBucket.Spec.SecretRef)
+	if err != nil {
+		log.Error(err, "Failed to get backup secret", "secret", client.ObjectKey{Namespace: backupBucket.Spec.SecretRef.Namespace, Name: backupBucket.Spec.SecretRef.Name})
+		r.Recorder.Eventf(backupBucket, corev1.EventTypeWarning, gardencorev1beta1.EventReconcileError, "Failed to get backup secret %s/%s: %w", backupBucket.Spec.SecretRef.Namespace, backupBucket.Spec.SecretRef.Name, err)
+		return reconcile.Result{}, err
+	}
+
+	if !controllerutil.ContainsFinalizer(secret, gardencorev1beta1.ExternalGardenerName) {
+		log.Info("Adding finalizer to secret", "secret", client.ObjectKeyFromObject(secret))
+		if err := controllerutils.AddFinalizers(ctx, r.GardenClient, secret, gardencorev1beta1.ExternalGardenerName); err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to add finalizer to secret: %w", err)
+		}
+	}
+
+	var (
+		reconcileExtensionBackupBucket    = false
+		lastErrorPresent                  = false
+		extensionSecret                   = r.emptyExtensionSecret(backupBucket.Name)
+		expectedExtensionBackupBucketSpec = extensionsv1alpha1.BackupBucketSpec{
+			DefaultSpec: extensionsv1alpha1.DefaultSpec{
+				Type:           backupBucket.Spec.Provider.Type,
+				ProviderConfig: backupBucket.Spec.ProviderConfig,
+			},
+			Region: backupBucket.Spec.Provider.Region,
+			SecretRef: corev1.SecretReference{
+				Name:      extensionSecret.Name,
+				Namespace: extensionSecret.Namespace,
+			},
+		}
+	)
+
+	if err := r.SeedClient.Get(ctx, client.ObjectKeyFromObject(extensionSecret), extensionSecret); err == nil {
+		if !reflect.DeepEqual(extensionSecret.Data, secret.Data) {
+			reconcileExtensionBackupBucket = true
+		}
+	}
+
+	if err := r.deployBackupBucketExtensionSecret(ctx, backupBucket); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err := r.SeedClient.Get(ctx, client.ObjectKeyFromObject(extensionBackupBucket), extensionBackupBucket); err != nil ||
+		!reflect.DeepEqual(extensionBackupBucket.Spec, expectedExtensionBackupBucketSpec) {
+		reconcileExtensionBackupBucket = true
+	} else {
+		var (
+			lastObservedError error
+			lastError         = extensionBackupBucket.Status.LastError
+			lastOperation     = extensionBackupBucket.Status.LastOperation
+		)
+
+		if lastError != nil || lastOperation == nil || lastOperation.State != gardencorev1beta1.LastOperationStateSucceeded {
+			reconcileExtensionBackupBucket = true
+			lastErrorPresent = true
```

Comment:
This is now set to true when `lastOperation` is `nil`? 😕

Reply:
Probably the variable name is confusing. Better call it `reconciliationSuccessful`?